Rebase of previous fixes

FraMecca opened the pull request:
As discussed by email, this rebase includes the fixes of the past month.

I want to be clear that tests and benchmarks run successfully but I would prefer to discuss them after merging with upstream.
The master branch has advanced quite a bit and some programs such as dustmite have changed parts of their code so newer versions don't even compile.

Commit:
options for concurrent behaviour

now concurrent behaviour can be choosed at runtime. Default to false.

shouldFork

Commit:
Allocate markbits as map_shared

When forking concurrently the markbits set by the children should be visible by the parent.

Commit:
setAll for gcbits and coherent function signatures

- setAll() is used to set all bits, in the same way it is used in tangort
- given that now many functions are called inside nogc functions (from the gcx class) gcbits' function signatures need to be marked as such

Commit:
concurrently set for markbits and freebits

Some pages might get updated or new pools can get created while the children is concurrently marking memory. This commit introduces the necessary sets and tests (for markbits and freebits) in order to avoid unnecessary scans or uncorrect marks.

Commit:
reflect functions in gcx from main class

Given that GC is a singleton that can be called from a program by a user,
I updated the signature of the function to accept a boolean variable that disables the forking behaviour.
It is defaulted to true so that the functions respect the behaviour of upstream.

Commit:
stats functions from tangorc

implements profiling for the gcx class.
At program termination it outputs a file "collect.stats" with informations about pause times.

Commit:
stats.d file from tangorc

this is needed for benchmarking the pause time.
Upstream has a different strategy for profiling and I think it will not get accepted.
In any case, it is very needed for profiling the gc at this stage

Commit:
helper functions for stats.d

This functions are needed by stats.d to gather information about the GC current state.

Commit:
added stats.d to SRCS

SRCS contains a list of files that gets compiled with the current makefile

llucax commented on src/gc/bits.d:
```diff
@@ -31,24 +31,27 @@ struct GCBits
+        import gc.os : os_mem_map;
         this.nbits = nbits;
-        data = cast(typeof(data[0])*)calloc(nwords, data[0].sizeof);
+        data = cast(typeof(data[0])*)os_mem_map(nwords * data[0].sizeof, true); // Allocate as MAP_SHARED
```
I think the memory should only be shared for mark bits, all the rest is better to keep them private to the process. In tangort there is an option to allocate the bits as shared or private memory.

llucax commented on src/gc/bits.d:
```diff
@@ -31,24 +31,27 @@ struct GCBits
     size_t nbits;
 
-    void Dtor() nothrow
+    void Dtor() nothrow @nogc
```
Not a big deal, but these `@nogc` are actually unrelated to this commit, right?

llucax commented on src/gc/os.d:
```diff
@@ -146,15 +146,16 @@ else static if (is(typeof(mmap)))  // else version (GC_Use_Alloc_MMap)
     enum { HAVE_FORK = true }
 
-    void *os_mem_map(size_t nbytes) nothrow
+    void *os_mem_map(size_t nbytes, bool share = false) nothrow
```
I prefer the more explicit approach used in tangort using an enum, but this is just bikeshedding :)

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2149,7 +2155,7 @@ struct Gcx
             pool = pooltable[n];
             pool.mark.zero();
-            if(!pool.isLargeObject) pool.freebits.zero();
+            // if (!pool.isLargeObject) pool.freebits.zero();
```
This is commented because now you keep them updated at all times, right?

FraMecca replied:
yep, this was actually suggested by you on email

llucax commented on src/gc/stats.d:
```diff
@@ -0,0 +1,439 @@
+ * Authors:   Walter Bright, Sean Kelly, Leandro Lucarella
+ */
+
+module gc.stats;
```
There is some overlap between the current stats collected by the GC and this, so I wonder how upstream will take this. But it proved to be very helpful in tangort.

FraMecca replied:
I think that upstream will refuse this but it is needed for our benchmarking and discussion.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2557,6 +2559,8 @@ Lmark:
             }
 
+            if(shouldFork)
+                thread_suspendAll();
```
Threads are not suspended if it's not forking?

Also, I don't get how this is related to `minimize()`.

llucax reviewed (commented):
This seems much cleaner than the previous PR, but still I find the commit message too terse. Sometimes if very hard to review a commit when the intention of what's going on is not stated clearly in the commit message (not only title). For example I have no idea about what's going on in commit `more concurrent behaviour with respect to minimize()` with `shouldFork`.

FraMecca commented:
Thank you Leandro.

If you prefer I can close the pull and rebase adding better commit messages.

If it is enough for you I'll explain the commits here:

 * [added stats.d to SRCS](https://github.com/FraMecca/druntime/pull/3/commits/023fec0b42d3d8f42a3707894a9d91fc9c56dd37): SRCS are used by the runtime to list the files to be compiled.

* [stats.d from tangort](https://github.com/FraMecca/druntime/pull/3/commits/e649d812ccd2e34caf1dc5784c3e685bf34871fb): taken from tangort, needs some functions to work properly and they are added in the other commits.

* [more concurrent behaviour with respect to minimize()](https://github.com/FraMecca/druntime/pull/3/commits/1a04181808d4a7f389b34f99bafb306c69b0c9d2): minimize() should be called after a fullcollect. For this reason I set a boolean shouldMinimize that is checked at the end of the fullcollect function.

* [options for concurrent behaviour](https://github.com/FraMecca/druntime/pull/3/commits/6bff6e3aadf420c8613381cc1b99cf1785103e21): we discussed this by email for many weeks. This should cover all the cases where allocations are being done and changes should be reflected on markbits and freebits.
 
* [reflect functions in gcx from main class](https://github.com/FraMecca/druntime/pull/3/commits/f685afa3c58efbd73b01913a2ca896f2f680bd35): the main GC class functions now have an option for forking that is defaulted to false.

* [options for concurrent behaviour](https://github.com/FraMecca/druntime/pull/3/commits/6bb22be594682e6650a1437fa7ccbb8a4a1d7a5d): forking behaviour can be enabled by a commandline switch.

* [allocate gcbits as mmap shared](https://github.com/FraMecca/druntime/pull/3/commits/89f8b4e16ed50d80b9cfa6e5c89d22a57453a8b7): as discussed by email, needed in order for the children to work correctly and notify its changes to the parent.

llucax commented:
I think is a good idea to always write more comprehensive commit messages in general, it really eases the process of reviewing code. In particular if you want to merge this upstream later on, it won't be a wasted effort.

You don't need to close this PR, you can rebase and force-push.

Thanks!

Commit:
Solved deadlock

The deadlock happened because when not forking thread_suspendAll() was called twice.
This is correct because when not forking threads are suspended above at line 2552

Commit:
Minimize() concurrently

druntime has a function for returning unused memory pools to the
OS ([minimize()] (https://dlang.org/phobos/core_memory.html#.GC.minimize)).
The function needs to be called after a fullcollect(). For this reason I used a boolean that is checked at the end of the fullcollect function.

Commit:
Merge branch 'rebase2' into cdgc

FraMecca commented:
Hi Leandro, 
I have branched, rebased and merged.

Looking forward to your review.

llucax commented on src/gc/bits.d:
```diff
@@ -27,28 +28,36 @@ struct GCBits
     enum BITS_SHIFT = (wordtype.sizeof == 8 ? 6 : 5);
     enum BITS_MASK = (BITS_PER_WORD - 1);
     enum BITS_1 = cast(wordtype)1;
+    bool mmap;
```
Why do you need this bool?

llucax commented on src/gc/bits.d:
```diff
@@ -27,28 +28,36 @@ struct GCBits
         {
-            free(data);
+            if (mmap)
+                os_mem_unmap(data, nwords * data[0].sizeof);
```
Ohhh. Why don't you just use mmap for everything instead? Is it to save memory because usually bitmaps are less than 4k?

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -1866,6 +1872,8 @@ struct Gcx
         {
             pool.initialize(npages, isLargeObject);
+            if (collectInProgress)
+                pool.mark.setAll();
```
Why was this moved here?

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -1131,12 +1131,24 @@ class ConservativeGC : GC
+    }
+
+
     void collect() nothrow
```
Why not just extending this one? Aren't both conflicting when `collect()` is called since the new function has a default value?

llucax reviewed (commented):
Thanks, that helped to understand some of the changes! Besides the small comments LGTM in general.
